File: 322.coin-change.py
# @leet start
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0

        memo = [0 for _ in range(amount + 1)]
        memo[0] = 0

        for i in range(1, amount + 1):
            local = sys.maxsize
            for coin in coins:
                if i - coin >= 0:
                    local = min(local, memo[i - coin] + 1)
            memo[i] = local

        return -1 if memo[-1] == sys.maxsize else memo[-1]

        
# @leet end

File: 322.coin-change.ts
// @leet start
function coinChange(coins: number[], amount: number): number {
	let memo: number[] = [0];
	let l: number = coins.length;

	for (let i: number = 1; i <= amount; i++) {
		let curr_val = -1;
		for (let j: number = 0; j < l; j++) {
			let val: number = coins[j];

			if (i - val >= 0 && memo[i-val] != -1) {
				if (curr_val === -1) {
					curr_val = memo[i-val] + 1;
				} else {
					curr_val = Math.min(curr_val, memo[i-val] + 1);
				}
			}
		}
		memo.push(curr_val)
	}

	return memo[amount];
    
};
// @leet end
